File: Week 6/sentimental-credit/credit.py
def get_card_number():
    while True:
        card_number = input('Number: ')

        if card_number.isnumeric():
            return int(card_number)


def get_card_parameters(card_number):
    sum_normal = 0
    sum_by_2 = 0
    sum_total = 0
    counter = 0

    while card_number > 0:
        counter += 1

        current_last_digit = card_number % 10

        if counter % 2 == 1:
            sum_normal += current_last_digit

        else:
            validator = 2 * current_last_digit
            if validator >= 10:
                sum_by_2 += 1 + (validator % 10)

            else:
                sum_by_2 += validator

        card_number = int(card_number / 10)

    sum_total = sum_normal + sum_by_2

    card_parameters = {'sum_total': sum_total, 'counter': counter}

    return card_parameters


def print_card_type(card_number, card_parameters):
    first_number = str(card_number)[0]
    second_number = str(card_number)[1]

    sum_total = card_parameters['sum_total']
    counter = card_parameters['counter']

    if sum_total % 10 != 0:
        print('INVALID')

    elif counter == 15 and first_number == '3' and (second_number == '3' or second_number == '7'):
        print('AMEX')

    elif counter == 13 or (counter == 16 and first_number == '4'):
        print('VISA')

    elif counter == 16 and first_number == '5' and (second_number >= '1' and second_number <= '5'):
        print('MASTERCARD')

    else:
        print('INVALID')


def main():
    card_number = get_card_number()
    card_parameters = get_card_parameters(card_number)
    print_card_type(card_number, card_parameters)


main()

File: Week 1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Número do cartão
    long number = get_long("Number: ");

    // Transformando em string para auxiliar mais tarde
    char card[20];
    sprintf(card, "%li", number);
    int firstNumber = (int) card[0] - 48, secondNumber = (int) card[1] - 48;

    // Verifica validade
    int sumNormal = 0, sumBy2 = 0, sumTotal = 0, counter = 0;

    // Pegar cada dígito a partir do último e somar à variável correspondente
    while (number > 0)
    {
        counter += 1;

        if (counter % 2 == 1)
        {
            sumNormal += number % 10;
        }
        else
        {
            int temp = 2 * (number % 10);
            if (temp >= 10)
            {
                sumBy2 += 1 + (temp % 10);
            }
            else
            {
                sumBy2 += temp;
            }
        }

        number /= 10;
    }

    // Efetuando a soma total
    sumTotal = sumNormal + sumBy2;

    // Verificando se o cartão é válido, se sim, qual seu tipo
    if (sumTotal % 10 != 0)
    {
        printf("INVALID\n");
    }
    else if (counter == 15 && firstNumber == 3 && (secondNumber == 3 || secondNumber == 7))
    {
        printf("AMEX\n");
    }
    else if (counter == 13 || (counter == 16 && card[0] == '4'))
    {
        printf("VISA\n");
    }
    else if (counter == 16 && firstNumber == 5 && (secondNumber >= 1 && secondNumber <= 5))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}